lang/cs/Org.Apache.REEF.IMRU.Tests/TestSystemStates.cs

```diff
@@ -0,0 +1,135 @@
+        }
+
+        [Fact]
+        public void TestNoRecoverableFailedEvaluatorDuringWaitingForEvaluator()
+        {
+            var systemState = new SystemStateMachine();
+            Assert.True(systemState.MoveNext(SystemStateEvent.NotRecoverable).Equals(SystemState.Fail), "Fail to move from WaitingForEvaluator state to Fail state.");
+        }
+
+        /// <summary>
+        /// Recovery scenario, from WaitingForEvaluator->SubmittingTasks->TasksRunning->ShuttingDown->WaitingForEvaluator
+        /// </summary>
+        [Fact]
+        public void TestFailedEvaluatorAfterTasksAreRunning()
+        {
+            var systemState = new SystemStateMachine();
+            systemState.MoveNext(SystemStateEvent.AllContextsAreReady);
+            systemState.MoveNext(SystemStateEvent.AllTasksAreRunning);
+            Assert.True(systemState.MoveNext(SystemStateEvent.FailedNode).Equals(SystemState.ShuttingDown), "Fail to move from TasksRunning state to ShuttingDown state.");
+            Assert.True(systemState.MoveNext(SystemStateEvent.Recover).Equals(SystemState.WaitingForEvaluator), "Fail to move from ShuttingDown state to WaitingForEvaluator state.");
+        }
+
+        /// <summary>
+        /// Recovery scenario, from WaitingForEvaluator->SubmittingTasks->TasksRunning->ShuttingDown->WaitingForEvaluator
+        /// </summary>
+        [Fact]
+        public void TestFailedTaskAfterTasksAreRunning()
+        {
+            var systemState = new SystemStateMachine();
+            systemState.MoveNext(SystemStateEvent.AllContextsAreReady);
+            systemState.MoveNext(SystemStateEvent.AllTasksAreRunning);
+            Assert.True(systemState.MoveNext(SystemStateEvent.FailedNode).Equals(SystemState.ShuttingDown), "Fail to move from TasksRunning state to ShuttingDown state.");
+            Assert.True(systemState.MoveNext(SystemStateEvent.Recover).Equals(SystemState.WaitingForEvaluator), "Fail to move from ShuttingDown state to WaitingForEvaluator state.");
+        }
+
+        /// <summary>
+        /// Fail scenario, from WaitingForEvaluator->SubmittingTasks->ShuttingDown->Fail
+        /// </summary>
+        [Fact]
+        public void TestFailedTaskDuringTaskSubmittingThenFailToRecovery()
+        {
+            var systemState = new SystemStateMachine();
+            systemState.MoveNext(SystemStateEvent.AllContextsAreReady);
+            Assert.True(systemState.MoveNext(SystemStateEvent.FailedNode).Equals(SystemState.ShuttingDown), "Fail to move from SubmittingTasks state to ShuttingDown state.");
+            Assert.True(systemState.MoveNext(SystemStateEvent.NotRecoverable).Equals(SystemState.Fail), "Fail to move from ShuttingDown state to Fail state.");
+        }
+
+        /// <summary>
+        /// Fail scenario, from WaitingForEvaluator->SubmittingTasks->ShuttingDown->WaitingForEvaluator
```

Comment:
If we keep this test of the identical pair, please update comment, it's not fail scenario but recovery.